### ISSUE TITLE
Container based executors - Artifacts

### ISSUE THREAD

**Issue by exu** [kubeshop/testkube#2133]
linked: https://github.com/kubeshop/testkube/issues/2061

- Allow users to define an executor by specifying 
         - docker image name
         - credentials (in case of private registries)
         - name of the executor (and other misc data t.b.d)
- Test can be defined to use this executor. The same behaviour is expected from a users's standpoint 
- There should be a clear folder structure documented that specifies
      - Which folder should the artifacts be saved too ?
      - Q: should the logs be saved in artifacts ?

- Pass & fail is solely determined by the exit status of the pods execution. but allow a Junit.xml report to be retrieved for extra info.
- Variables of type basic and secret should be mounted as environment variables on the container.
- Arguments defined by the user should be sent in a similar faction as this example:
    - if the user were to run a nginx image in k8s and send the command `ls –la /` in k8s, he would do it like this -> `kubectl exec nginx -- ls –la /`.  A similar behaviour is expected for the this container based executor and its arguments.
- Are resource limit important to set ? ``resources:
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "128Mi"
        cpu: "500m" ``

**Comment by TheBrunoLopes** [comment]
@ili-shmidt-wib here's the description of the feature we were discussing last week.
Feel free to add your input 👌

**Comment by TheLeanMachine** [comment]
Hey @TheBrunoLopes ,

hate to ask, but: Is there a rough(!) ETA on this feature?

(I seriously don't want anyone to rush, but I will have to make a decision: Should I wait for this feature ...OR... seek out an try to find alternative for running my tests)

**Comment by TheLeanMachine** [comment]
btw: @exu , @povilasv 

...as soon as this is released - or some kind of beta-version is publicly available -, I would our this "generic container executor" out.

**Comment by TheBrunoLopes** [comment]
Hi @TheLeanMachine it's awesome that you are also wanting to test this feature out!
- The images don't need to be public, you will be able to define the pull secret of this image in case it's not already configured in your k8s cluster.
- We started implementing it yesterday, I can't promise you an exact date yet, but I would venture saying that it will probably take ~2 weeks to make a release out of it. 

- Is there a specific type of test that you want to run ? maybe we can help you with that.

**Comment by TheLeanMachine** [comment]
Hey @TheBrunoLopes ,

(first of all: Thanks 4 the quick and conprehensive reply)

- Cool, will learn about "pull secrets & friends", then
- Awesome! - It's really only about knowing that it shouldn't be 2 months or 2 years ;-)
- Yes, indeed. I still have the problem with running Cypress tests that [are aggregated using NX](https://nx.dev/packages/cypress) and "good old" Java-API-tests implemented with Maven (that fails using testkube's Maven runner - due to some Java version problems)
   - I thought, I tried out both using the "Generic Image Runnter" (or how do you wanna call it)
   - From my "naive end-user viewoint", I imagine the existing custom-runners (Maven, Cypress, ...) to delegate to the "Generic Image Runner". But - of course - I lack implementation details, so maybe that's not feasible

**Comment by exu** [comment]
@TheLeanMachine please pass those Java problems in separate issue we can try to help with that.

**Comment by TheLeanMachine** [comment]
Thanks @exu for reaching out,

in fact, when I was adding the JDK-18-executor [following the corresponding documentation](https://kubeshop.github.io/testkube/test-types/executor-maven/) (section "Using different JDKs"), it worked just fine.

**Comment by TheLeanMachine** [comment]
Hey @povilasv ,

awesome!
Will definitely have a look at the docs and checkout ASAP, **but**...

...ATM, I am sadly blocked by this bug: https://github.com/kubeshop/testkube/issues/2361
 ;-)

After this is resolved and the latest testkube release is deployed, I will jump on this and will feed back the results.

**Comment by TheLeanMachine** [comment]
Hey @povilasv ,

I know that it's not the right thread here, but:
Please don't make the mistake of CI/CD-providers - like bitbucket - to only obtain the artefacts in case of build success... ;-)

**Comment by TheBrunoLopes** [comment]
Thanks for the feedback @TheLeanMachine. We will be aware of that.
@povilasv noticed that #2361 is closed. Is it fixed ?

**Comment by TheLeanMachine** [comment]
to @TheBrunoLopes and @povilasv 

I am not 100% sure, if this is still a problem. Maybe, I simply have to install a more recent testkube client.
Will do that tomorrow and update you, here.

**Comment by TheLeanMachine** [comment]
Hi @povilasv , @TheBrunoLopes ,

I dunno, if https://github.com/kubeshop/testkube/issues/2361 is still a problem...

At least I, personally, was able to open the testkube dashboard on my MacBook again after _using testkube client from kubeshop's tap_...

```
(brew uninstall testkube) // removing old client
brew tap kubeshop/homebrew-testkube
brew install kubeshop/testkube/testkube
```

**Comment by TheLeanMachine** [comment]
Hey @povilasv,

(using these versions, now)
```
Client Version 1.5.41
Server Version v1.5.41
```

here some initial feedback: I created my own custom test - still [want to run NX-agregated tests](https://github.com/kubeshop/testkube/issues/2141) - using an official NX-image, now. This is the executor CRD:
```
apiVersion: executor.testkube.io/v1
kind: Executor
metadata:
  name: nx-container-executor
  namespace: testkube
spec:
  image: soleware/nx-cli:8.5.2
  command: ["nx"]
  executor_type: container
  types:
  - nx-container/test
```

...with which I continue to create this test...
```
kubectl testkube create test --name nx-containerexecutor-test-v1 --type nx-container/test --executor-args '--version' --git-branch master --git-uri https://bitbucket.org/<MY_ORG>/<NX_REPO>.git --git-username kai_hoelscher --git-token <SECRET_TOKEN>
```

Running the test yields this output (I expected it to print the NX-version)...
![Screenshot 2022-09-30 at 16 49 40](https://user-images.githubusercontent.com/3470837/193296698-262cc18c-599a-4165-bba5-3ebcfb208589.png)

...which made me realise two things:

1. Apparently no git-checkout is performed
2. I forgot to ask for sth like a `pre-command-build-hook` (a command used to build/bundle the tests, before execution: here: `npm install`)

**Comment by TheLeanMachine** [comment]
Hey @povilasv,

sounds reasonable. I will check that out, soon...

**Comment by TheLeanMachine** [comment]
Hey @povilasv,

it seems, that - in fact - _I will be able to "hack multiple commands together'_. One way to implement this, is using a startup-command in the executor-definition (see `command` below) like this...

```
apiVersion: executor.testkube.io/v1
kind: Executor
metadata:
  name: nx-custom-executor
  namespace: testkube
spec:
  image: soleware/nx-cli:8.5.2
  command: ["/bin/bash", "-c", "pwd;echo 'Navigate to checked-out repository located in /data/repo';cd /data/repo; npm install"]
  executor_type: container
  types:
  - nx-custom/test
```

...but sadly that leads me to the next problem: Non writable file-system!
...so, when I create a testkube test using the above executor, I get this output:

![Screenshot 2022-10-07 at 15 48 45](https://user-images.githubusercontent.com/3470837/194569383-c596421e-6159-47e2-a656-160335138797.png)

_Any idea, how I could address this?_ (spoiler: `chmod -R 777 ...` did not help...)

**Comment by TheLeanMachine** [comment]
Hey @povilasv ,

thanks for your feedback, so far. Sounds reasonalble. Will think about this again...

(but basically, you're right)

**Comment by vsukhin** [comment]
So,remained part for this ticket is a storing artifact to external folder

**Comment by TheBrunoLopes** [comment]
AC:
- There should be an environment variable that is possible to change/configure that specifies the dir path to where to retrieve the artifacts.
